cms/static/js/views/pages/container.js

```diff
@@ -607,6 +609,45 @@ function($, _, Backbone, gettext, BasePage,
             });
         },
 
+        pickLibraryComponent: function(template, target) {
+            // A placeholder element is created in the correct location for the new xblock
+            // and then onNewXBlock will replace it with a rendering of the xblock. Note that
+            // for xblocks that can't be replaced inline, the entire parent will be refreshed.
+            var parentElement = this.findXBlockElement(target),
+                parentLocator = parentElement.data('locator'),
+                buttonPanel = target.closest('.add-xblock-component'),
+                listPanel = buttonPanel.prev(),
+                scrollOffset = ViewUtils.getScrollOffset(buttonPanel),
+                $placeholderEl = $(this.createPlaceholderElement()),
+                requestData = _.extend(template, {
+                    parent_locator: parentLocator
+                }),
+                placeholderElement;
+            placeholderElement = $placeholderEl.appendTo(listPanel);
+            // Create an blank xblock with the same category as the upstream library xblock
+            return $.postJSON(
+                this.getURLRoot() + '/',
+                requestData,
+                function(data) {
+                    // Link with upstream library xblock and sync
+                    $.ajax({
+                        url: `/api/contentstore/v2/downstreams/${data.locator}`,
+                        type: 'PUT',
+                        dataType: 'json',
+                        contentType: 'application/json',
+                        data: JSON.stringify({
+                            upstream_ref: requestData.usageKey,
+                            sync: true,
+                        }),
+                        success: _.bind(this.onNewXBlock, this, placeholderElement, scrollOffset, false, data),
+                    });
+                }.bind(this)
+            ).fail(function() {
+                // Remove the placeholder if the update failed
+                placeholderElement.remove();
+            });
+        },
+
         duplicateComponent: function(xblockElement) {
             // A placeholder element is created in the correct location for the duplicate xblock
             // and then onNewXBlock will replace it with a rendering of the xblock. Note that
```

Comment:
This file has many duplicate functions like `deleteXBlock`, `duplicateXBlock`, `showMoveXBlockModal` etc.